### PR TITLE
Drawdown control w/ user-defined Severity & Penalty Levels

### PR DESCRIPTION
**Pull request by wilhempujar**
This patch provides an alternate, user-defined reward calculation approach vs the default deviation formula that we use in the SLO registry. 

It allows the contract creator to define severity levels, and their respective penalties.
If the SLI is equal or greater than the value of a severity level `severity[i]`, then `penalty[i]` becomes the deviation.

N.B. In DSLA v2.0, deviation = reward.

In a Parametric Staking SLA, this gives the contract creator a way to control the drawdown of the liquidity pool based on the price increase of the underlying asset. 

For the DSLA token initial configuration:
* **Severity 1**: Any increase under 100% = 1% deviation
* **Severity 2**: An increase of 100%+ = 10% deviation
* **Severity 3**: An increase of 250%+ = 25% deviation

25% is also the protocol-level cap on rewards.

The DTK needs to be updated accordingly (I'm on it).

### CODE REVIEW

File: contracts/SLORegistry.sol [hunk 1]
```diff
@@ -108,27 +108,40 @@ contract SLORegistry {
...
-        // The deviation is the percentage difference between SLI and SLO
-        //                          | sloValue - sli |
-        // formula =>  deviation = -------------------- %
-        //                          (sli + sloValue) / 2
-        uint256 deviation = ((
-            _sli >= sloValue ? _sli - sloValue : sloValue - _sli
-        ) * 20000) / (_sli + sloValue);
+        uint256 deviation = 0;
+
+        for (uint256 i = 0; i < severity.length; i++) {
```
Comment (Axel-CH) [inline contracts/SLORegistry.sol:122]: What append if the arrays don't have the same length? Should we add a check to validate that they have?

File: contracts/SLORegistry.sol [hunk 2]
```diff
@@ -108,27 +108,40 @@ contract SLORegistry {
...
+            if (_sli >= uint256(severity[i])) {
+                deviation = uint256(penalty[i]);
+            }
+        }
+
+        if (deviation == 0) {
```
Comment (Axel-CH) [inline contracts/SLORegistry.sol:128]: If I have created two arrays of penalty / severity but for some reason at one time the deviation is equal to 0.
I will fall back to the previous formula and my custom penalty won't be respected.

I think we need to replace this with a check of presence / length of penalty / severity arrays

**Review by Axel-CH — CHANGES_REQUESTED** [review]
Great changes that will enable more customization on SLA outcome, need to update tests with new createSLA params

File: contracts/SLORegistry.sol (hunk 2 above)
Comment (wilhempujar) [reply contracts/SLORegistry.sol:128]: If the deviation equals 0, meaning `penalty[i] = 0`, we can assume that's what the user wants imho. It makes a nice fallback to the default behaviour, when using custom severities.

**Comment by wilhempujar** [comment]
> Great changes that will enable more customization on SLA outcome, need to update tests with new createSLA params

Tests updated / Passing ☑️

LFG?

File: contracts/SLORegistry.sol (hunk 2 above)
Comment (Axel-CH) [reply contracts/SLORegistry.sol:128]: Are we sure that the deviation will stay at zero after the fallback to default formula, as this formula is based on the SLO / SLI delta?

**Comment by Axel-CH** [comment]
Specific tests for user defined severity & penalty levels should be added imo, beside that great work. LGTM :+1:

**Review by Axel-CH — APPROVED** [review]

**Pull request merged**